staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filter-config.go

```diff
@@ -0,0 +1,361 @@
+			plState.queues.SetConfiguration(plState.qsConfig) // TODO: make sure that retains other config if zero desired queues
+		}
+	}
+	reqMgr.curState.Store(newRMState)
+	klog.V(5).Infof("Switched to new RequestManagementState")
+	// We do the following only after updating curState to guarantee
+	// that if Wait returns `tryAnother==true` then a fresh load from
+	// curState will yield an requestManagerState that is at least
+	// as up-to-date as the data here.
+	//
+	// Put another way: doing the Quiesce calls after Store(X) ensures
+	// that if any call to qs.Quiesce causes a call to qs.Wait to
+	// return with `tryAnother==true` then a following call to Load
+	// will return X or an even more recently stored value.
+	for _, plState := range newlyQuiescent {
+		plState.queues.Quiesce(plState.emptyHandler)
+	}
+}
+
+func qscOfPL(pl *rmtypesv1a1.PriorityLevelConfiguration, requestWaitLimit time.Duration) (fq.QueueSetConfig, error) {
+	qsConfig := fq.QueueSetConfig{Name: pl.Name,
+		RequestWaitLimit: requestWaitLimit}
+	var err error
+	if qc := pl.Spec.Limited.LimitResponse.Queuing; qc != nil {
+		dealer, e1 := shufflesharding.NewDealer(int(qc.Queues), int(qc.HandSize))
+		if e1 != nil {
+			err = errors.Wrap(e1, fmt.Sprintf("priority level %q has QueuingConfiguration %#+v, which caused shufflesharding.NewDealer to fail", pl.Name, *qc))
+		}
+		qsConfig.DesiredNumQueues = int(qc.Queues)
+		qsConfig.QueueLengthLimit = int(qc.QueueLengthLimit)
+		qsConfig.Dealer = dealer
+	}
+	return qsConfig, err
+}
+
+func (reqMgr *requestManager) syncFlowSchemaStatus(fs *rmtypesv1a1.FlowSchema, isDangling bool) {
+	danglingCondition := apihelpers.GetFlowSchemaConditionByType(fs, rmtypesv1a1.FlowSchemaConditionDangling)
+	if danglingCondition == nil {
+		danglingCondition = &rmtypesv1a1.FlowSchemaCondition{
+			Type: rmtypesv1a1.FlowSchemaConditionDangling,
+		}
+	}
+
+	switch {
+	case isDangling && danglingCondition.Status != rmtypesv1a1.ConditionTrue:
+		danglingCondition.Status = rmtypesv1a1.ConditionTrue
+		danglingCondition.LastTransitionTime = metav1.Now()
+	case !isDangling && danglingCondition.Status != rmtypesv1a1.ConditionFalse:
+		danglingCondition.Status = rmtypesv1a1.ConditionFalse
+		danglingCondition.LastTransitionTime = metav1.Now()
+	default:
+		// the dangling status is already in sync, skip updating
+		return
+	}
+
+	apihelpers.SetFlowSchemaCondition(fs, *danglingCondition)
+
+	_, err := reqMgr.flowcontrolClient.FlowSchemas().UpdateStatus(fs)
+	if err != nil {
+		klog.Warningf("failed updating condition for flow-schema %s", fs.Name)
+	}
+}
+
+func (newRMState *requestManagerState) imaginePL(proto *rmtypesv1a1.PriorityLevelConfiguration, requestWaitLimit time.Duration, shareSum *float64) {
```

Comment:
imagining from the proto-PL can be fragile b/c it doesnt go thru defaulting

Reply:
Yes, I am going to refactor so that the typed objects have defaulting applied.

Reply:
No, dang, I can't --- because the default objects are provided by k8s.io/apiserver, which can not depend on the stuff the does the defaulting.  I will have to make the config controller copy and apply defaulting.

Reply:
dang, the config controller is in k8s.io/apiserver too!

Reply:
I think I suggested in another comment just adding a test to ensure that defaulting has no effect on these objects. The test can live somewhere that can easily import all the needed things.

Reply:
Yes, @lavalamp , that looks like the easiest approach.  We do not really need the option for the bootstrap package to supply objects that defaulting would change.  I am pursuing that approach.

Reply:
pushed yesterday